xdsl/transforms/stencil_inlining.py

```diff
@@ -0,0 +1,368 @@
+)
+from xdsl.rewriter import InsertPoint
+from xdsl.transforms.canonicalization_patterns.stencil import (
+    ApplyRedundantOperands,
+    ApplyUnusedOperands,
+    ApplyUnusedResults,
+)
+from xdsl.transforms.shape_inference_patterns.stencil import update_result_size
+from xdsl.transforms.stencil_unroll import offseted_block_clone
+
+
+def is_before_in_block(op1: Operation, op2: Operation):
+    """
+    Check if op1 is before op2 in the same block.
+    """
+    block = op1.parent
+    assert block is not None
+    assert block is op2.parent
+    return block.get_operation_index(op1) < block.get_operation_index(op2)
+
+
+class StencilStoreResultForwardPattern(RewritePattern):
+    """
+    Replace non-empty `stencil.store_result`s by their argument.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: StoreResultOp, rewriter: PatternRewriter, /):
+        if op.arg is None:
+            return
+        rewriter.replace_matched_op([], [op.arg])
+
+
+class StencilIfResultForwardPattern(RewritePattern):
+    """
+    Replace `!stencil.result<T>`-typed scf.if by `T`-typed scf.if.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: scf.IfOp, rewriter: PatternRewriter, /):
+        result_types = [r.type for r in op.output]
+        new_result_types = [
+            t.elem if isinstance(t, ResultType) else t for t in result_types
+        ]
+        if new_result_types == result_types:
+            return
+        rewriter.replace_matched_op(
+            scf.IfOp(
+                op.cond,
+                new_result_types,
+                op.detach_region(0),
+                op.detach_region(0),
+            )
+        )
+
+
+def has_single_consumer(producer: ApplyOp, consumer: ApplyOp):
+    """
+    Check if the producer has a single consumer.
+    """
+    return all(
+        isinstance(u.operation, ApplyOp) and u.operation == consumer
+        for r in producer.results
+        for u in r.uses
+    )
+
+
+def is_rerouting_possible(producer: ApplyOp, consumer: ApplyOp):
+    """
+    Check if rerouting is possible.
+    """
+    # Perform producer consumer inlining instead
+    if has_single_consumer(producer, consumer):
+        return False
+    for operand in consumer.operands:
+        if isinstance(operand.owner, Operation):
+            if (operand.owner is not producer) and is_before_in_block(
+                producer, operand.owner
+            ):
+                return False
+    return True
+
+
+def is_inlining_possible(producer: ApplyOp, consumer: ApplyOp):
+    """
+    Check if inlining is possible.
+    """
+    # Don't inline any producer with conditional writes.
+    r = not any(
+        store_result.arg is None
+        for store_result in producer.walk()
+        if isinstance(store_result, StoreResultOp)
+    ) and not any(
+        # Don't inline any dynamic accesses.
+        isinstance(use.operation, DynAccessOp)
+        for consumer_operand in consumer.operands
```

Comment:
nit:
```suggestion
        for consumer_operand, arg in zip(consumer.operands, consumer.region.block.args, strict=True)
```

Reply:
I think the code is doing a lookup rather than a zip at this point.